File: app/src/main/java/com/arjanvlek/oxygenupdater/internal/FunctionalAsyncTask.java
package com.arjanvlek.oxygenupdater.internal;

import android.os.AsyncTask;

import java8.util.function.Consumer;
import java8.util.function.Function;

/**
 * Oxygen Updater - © 2017 Arjan Vlek
 */

@SuppressWarnings("WeakerAccess")
public class FunctionalAsyncTask<Params, Progress, Result> extends AsyncTask<Params, Progress, Result> {

	private final Worker preExecuteFunction;
	private final Function<Params[], Result> backgroundFunction;
	private final Consumer<Progress[]> progressUpdateFunction;
	private final Consumer<Result> postExecuteFunction;

	public FunctionalAsyncTask(Worker preExecuteFunction, Function<Params[], Result> backgroundFunction, Consumer<Result> postExecuteFunction) {
		this(preExecuteFunction, backgroundFunction, postExecuteFunction, null);
	}

	public FunctionalAsyncTask(Worker preExecuteFunction, Function<Params[], Result> backgroundFunction, Consumer<Result> postExecuteFunction, Consumer<Progress[]> progressUpdateFunction) {
		this.postExecuteFunction = postExecuteFunction;
		this.preExecuteFunction = preExecuteFunction;
		this.backgroundFunction = backgroundFunction;
		this.progressUpdateFunction = progressUpdateFunction;
	}

	@Override
	protected Result doInBackground(Params[] params) {
		return backgroundFunction.apply(params);
	}

	@Override
	protected void onPreExecute() {
		preExecuteFunction.start();
	}

	@Override
	protected void onPostExecute(Result result) {
		postExecuteFunction.accept(result);
	}

	@SafeVarargs
	@Override
	protected final void onProgressUpdate(Progress... progress) {
		if (progressUpdateFunction != null) {
			progressUpdateFunction.accept(progress);
		}
	}
}

File: app/src/main/java/com/arjanvlek/oxygenupdater/internal/FunctionalAsyncTask.kt
package com.arjanvlek.oxygenupdater.internal

import android.os.AsyncTask

/**
 * @author Adhiraj Singh Chauhan (github.com/adhirajsinghchauhan)
 * @author Arjan Vlek (github.com/arjanvlek)
 */
class FunctionalAsyncTask<Params, Progress, Result> @JvmOverloads constructor(
    private val preExecuteFunction: () -> Unit,
    private val backgroundFunction: KotlinFunction<Array<Params>, Result>,
    private val postExecuteFunction: KotlinCallback<Result>,
    private val progressUpdateFunction: KotlinCallback<Array<Progress>>? = null
) : AsyncTask<Params, Progress, Result>() {

    override fun doInBackground(params: Array<Params>) = backgroundFunction.invoke(params)

    override fun onPreExecute() = preExecuteFunction.invoke()

    override fun onPostExecute(result: Result) = postExecuteFunction.invoke(result)

    override fun onProgressUpdate(progress: Array<Progress>) {
        progressUpdateFunction?.invoke(progress)
    }

}
